### PR TITLE
Feature 23 new graph library

### LINKED ISSUE #23: Changing graph library
As a developer I want to change graph library because current is buggy.

### PR DESCRIPTION
**Pull request by platonovsimeon**
resolves #23

### CODE REVIEW

File: src/components/index/Graph.tsx [hunk 1]
```diff
@@ -1,566 +1,436 @@
...
-  
-  /**
-   * Renders a node
-   */
-  private renderNode = (nodeRef: any, data: any, id: string, selected: boolean, hovered: boolean) => {
-    const props = {
-      height: 0,
-      width: 0
-    };
+  public GraphViewRef: any;
```
Comment (anttileppa) [inline src/components/index/Graph.tsx]: this should be lowercased, moved into top of the class and marked as private

File: src/components/index/Graph.tsx [hunk 2]
```diff
@@ -1,566 +1,436 @@
...
   }
-
-  /**
-   * Handles node selection
-   */
-  private onSelectNode = (viewNode: INode | null) => {
-    // Deselect events will send Null viewNode
-    this.setState({ selected: viewNode });
-    this.props.onSelectNode(viewNode);
+ // Handles node search
```
Comment (anttileppa) [inline src/components/index/Graph.tsx]: Change to proper tsdoc format

File: src/components/index/Graph.tsx [hunk 3]
```diff
@@ -1,566 +1,436 @@
...
+ private searchKnots = (): string[] => {
```
Comment (anttileppa) [inline src/components/index/Graph.tsx]: Fix indent (this and all following)

File: src/components/index/GraphView.tsx [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+  onNodeDragEnd: (viewNode: INode) => void;
+  onCreateNode: (viewNode: INode) => void;
+  onDeleteNode: (viewNode: INode) => Promise<void>;
+  onCreateEdge: (targetViewNode: INode, sourceViewNode: INode) => Promise<void>;
+  onDeleteEdge: (viewEdge: IEdge) => Promise<void>;
+  onEdgeClick: (viewEdge: IEdge) => void;
+  onUpdateMultiple: (viewNodes: INode[]) => Promise<void>;
+}
+interface IState {
+
```
Comment (anttileppa) [inline src/components/index/GraphView.tsx:37]: Remove empty line

File: src/actions/index.tsx [hunk 5]
```diff
@@ -1,125 +1,173 @@
...
 }
 
-export interface IntentDeleted {
-  type: constants.INTENT_DELETED,
-  intentId: string
+export interface IIntentDeleted {
+  type: constants.INTENT_DELETED;
+  intentId: string;
+}
+export interface ISetKnotPositions {
```
Comment (anttileppa) [inline src/actions/index.tsx:55]: There should be empty line between two interfaces

File: src/components/index/GraphView.tsx [hunk 6]
```diff
@@ -0,0 +1,443 @@
+import * as dagre from "dagre";
+import * as React from "react";
+import GraphConfig from "../../utils/graph-config";
+export interface INode {
```
Comment (anttileppa) [inline src/components/index/GraphView.tsx:5]: Empty line

File: src/components/index/GraphView.tsx [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+  id: string;
+  title: string;
+  x: number;
+  y: number;
+  type?: string;
+  subtype?: string;
+}
+export interface IEdge {
```
Comment (anttileppa) [inline src/components/index/GraphView.tsx:14]: There should be an empty line between interfaces

File: src/components/index/GraphView.tsx [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+  selectedEdge?: IEdge;
+  zoom: number;
+  translateX: number;
+  translateY: number;
+  filterIds: string[];
+  autolayout: boolean;
+}
+
+class GraphView extends React.Component<IProps, IState> {
+  // Get called when props change
```
Comment (anttileppa) [inline src/components/index/GraphView.tsx]: Proper tsdoc (change all)

File: src/components/index/GraphView.tsx [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+    return(
+      <div id="GraphView">
+      <svg
+      width={this.props.width}
+      onMouseMove={this.handleDrag}
+      onMouseUp={this.svgMouseUpHandler}
+      onClick={this.svgClickHandler}
+      onWheel={this.zoomHandler}
+      height={this.props.height}>
+
```
Comment (anttileppa) [inline src/components/index/GraphView.tsx]: Remove empty line

File: src/components/index/IntentEditor.tsx [hunk 10]
```diff
@@ -95,26 +95,40 @@ class IntentEditor extends React.Component<Props, State> {
         <Form inverted>
           <Form.Field>
             <label>Intent name</label>
-            <Input value={ this.state.intent.name } style={ { width: "100%" } } onChange={ this.onIntentNameChange } />
+            <Input
+            value={ this.state.intent.name }
```
Comment (anttileppa) [inline src/components/index/IntentEditor.tsx]: Indent attributes with two spaces

File: src/components/index/IntentEditor.tsx [hunk 11]
```diff
@@ -160,16 +190,17 @@ class IntentEditor extends React.Component<Props, State> {
     this.setState({loading: true});
 
     const intent = await Api.getIntentsService(this.props.keycloak ? this.props.keycloak.token! : "").findIntent(this.props.storyId, this.props.intentId);
-    
+
     this.setState({
+      intent,
       loading: false,
-      intent: intent
+
```
Comment (anttileppa) [inline src/components/index/IntentEditor.tsx]: Remove empty line

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: src/components/index/Graph.tsx [hunk 12]
```diff
@@ -1,566 +1,439 @@
...
   /**
    * Updates component state when knots or intents are changed
    */
-  static getDerivedStateFromProps = (props: Props, state: State) => {
-    const globalNode: INode = {
-      id: GLOBAL_NODE_ID,
-      title: "Global", // Localize
-      type: GLOBAL_TYPE
-    };
+  public static getDerivedStateFromProps = (props: IProps, state: IState) => {
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/components/index/Graph.tsx:59]: done

File: src/components/index/Graph.tsx [hunk 13]
```diff
@@ -1,566 +1,439 @@
...
   /**
    * Translates knot into node
-   * 
+   *
    * @param knot knot to translate
    * @param x x coordinate
    * @param y  y coordinate
    */
-  private static translateKnot(knot: Knot, x?: number, y?: number): INode {
+  private static translateKnot(knot: Knot, x: number, y: number): INode {
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/components/index/Graph.tsx:119]: done

File: src/components/index/Graph.tsx [hunk 14]
```diff
@@ -1,566 +1,439 @@
...
-
-    const defSvgNodeElement: any = nodeTypeXlinkHref ? document.querySelector(`defs>${nodeTypeXlinkHref}`) : null;
-    const nodeWidthAttr = defSvgNodeElement ? defSvgNodeElement.getAttribute('width') : 0;
-    const nodeHeightAttr = defSvgNodeElement ? defSvgNodeElement.getAttribute('height') : 0;
-    props.width = nodeWidthAttr ? parseInt(nodeWidthAttr, 10) : props.width;
-    props.height = nodeHeightAttr ? parseInt(nodeHeightAttr, 10) : props.height;
-    const index = this.props.knots.findIndex((knot) => {
-      return knot.id === id;
-    });
+  private static translateIntent(intent: Intent, sourceNode: INode, targetNode: INode): IEdge {
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/components/index/Graph.tsx:133]: done

File: src/components/index/IntentEditor.tsx (hunk 11 above)
Comment (platonovsimeon) [reply src/components/index/IntentEditor.tsx]: Empty line removed

File: src/components/index/GraphView.tsx (hunk 9 above)
Comment (platonovsimeon) [reply src/components/index/GraphView.tsx]: Empty line removed

File: src/components/index/GraphView.tsx (hunk 7 above)
Comment (platonovsimeon) [reply src/components/index/GraphView.tsx:14]: Added empty lines.

File: src/components/index/GraphView.tsx (hunk 6 above)
Comment (platonovsimeon) [reply src/components/index/GraphView.tsx:5]: Empty line added

File: src/actions/index.tsx (hunk 5 above)
Comment (platonovsimeon) [reply src/actions/index.tsx:55]: Added an empty line

File: src/components/index/IntentEditor.tsx (hunk 10 above)
Comment (platonovsimeon) [reply src/components/index/IntentEditor.tsx]: Indents added

File: src/components/index/GraphView.tsx (hunk 8 above)
Comment (platonovsimeon) [reply src/components/index/GraphView.tsx]: Formatted to tsdoc

File: src/components/index/Graph.tsx (hunk 1 above)
Comment (platonovsimeon) [reply src/components/index/Graph.tsx]: graphViewRef is not being used

File: src/components/index/Graph.tsx (hunk 2 above)
Comment (platonovsimeon) [reply src/components/index/Graph.tsx]: changed to tsdoc

File: src/components/index/GraphView.tsx (hunk 4 above)
Comment (platonovsimeon) [reply src/components/index/GraphView.tsx:37]: Removed

File: src/components/index/Graph.tsx (hunk 3 above)
Comment (platonovsimeon) [reply src/components/index/Graph.tsx]: fixed

File: src/components/index/Graph.tsx (hunk 1 above)
Comment (platonovsimeon) [reply src/components/index/Graph.tsx]: solved

File: src/components/index/Graph.tsx (hunk 2 above)
Comment (platonovsimeon) [reply src/components/index/Graph.tsx]: solved

File: src/components/index/Graph.tsx (hunk 3 above)
Comment (platonovsimeon) [reply src/components/index/Graph.tsx]: solved

File: src/components/index/GraphView.tsx (hunk 8 above)
Comment (platonovsimeon) [reply src/components/index/GraphView.tsx]: solved

File: src/components/index/IntentEditor.tsx (hunk 10 above)
Comment (platonovsimeon) [reply src/components/index/IntentEditor.tsx]: solved

File: src/components/index/GraphView.tsx (hunk 4 above)
Comment (platonovsimeon) [reply src/components/index/GraphView.tsx:37]: solved

**Pull request merged**